Source/bmalloc/libpas/src/libpas/pas_thread.c

```diff
@@ -0,0 +1,193 @@
+#include "pas_thread.h"
+#include "pas_utils.h"
+#include "pas_log.h"
+
+#include <windows.h>
+
+static const bool verbose = true;
+
+int pthread_create(pthread_t * tid,
+                   const pthread_attr_t * attr,
+                   unsigned (*start) (void *),
+                   void *arg)
+{
+    PAS_UNUSED_PARAM(tid);
+    PAS_UNUSED_PARAM(attr);
+    PAS_UNUSED_PARAM(arg);
+
+    // Create thread handle
+    HANDLE hThread;
+    
+    // Thread ID (optional)
+    unsigned threadIdentifier = 0;
+        
+    // Create the thread
+    hThread = (HANDLE)_beginthreadex(
+        NULL,
+        0,
+        start,
+        NULL,
+        0,
+        &threadIdentifier
+    );
+    
+    if (hThread == NULL) {
+        if (verbose)
+            pas_log("Failed to create thread.\n");
+        return 1;
+    }
+    
+    return 0;
+}			
+
+int pthread_detach(pthread_t thread) {
+    PAS_UNUSED_PARAM(thread);
+    // Detach is a no-op on Windows
+    return 0;
+}
+
+int pthread_getname_np(pthread_t thread, const char *name, size_t len) {
+    // This is only used for dump_thread_diagnostics
+    PAS_UNUSED_PARAM(thread);
+    PAS_UNUSED_PARAM(name);
+    PAS_UNUSED_PARAM(len);
+
+    PAS_ASSERT(false);
+
+    return 0;
+}
+
+pthread_t pthread_self(void) {
+    return GetCurrentThreadId();
+}
+
+int sched_yield() {
+    // SwitchToThread only yields to threads on the same processor
+    // If that fails we'll Sleep, which will also yield for threads ready to run on other processors
+    if (!SwitchToThread()) {
+        Sleep(0);
+    }
+    return 0;
+}
+
+// This is used to wrap the passed void(*)(void) function so it can be passed to InitOnceExecuteOnce
+BOOL once_init_runner(
+    PINIT_ONCE once_control,
+    PVOID init_routine,
+    PVOID *context)
+{
+    PAS_UNUSED_PARAM(once_control);
+    PAS_UNUSED_PARAM(context);
+    ((void (*)(void))init_routine)();
+
+    return TRUE;
+}
+
+int pthread_once(pthread_once_t *once_control, void (*init_routine)(void)) {
+    int result;
+    result = InitOnceExecuteOnce(*once_control,once_init_runner,init_routine,NULL);
+    if (verbose && !result) {
+        pas_log("Failed to run pthread_once.\n");
+    }
+
+    return result;
+}
+
+// Thread Local Storage
+// The regular Windows Thread Local Storage APIs like TlsAlloc don't have a way an easy way 
+// to register a per-thread destructor to be able to clean up the memory.
+// Blink gets around this with some crazy pragmas to manually insert a function to be called 
+// on each thread's exit:
+// https://web.archive.org/web/20070921204540/https://www.codeproject.com/threads/tls.asp
+// https://source.chromium.org/chromium/chromium/src/+/main:base/threading/thread_local_storage_win.cc;l=38-48
+//
+// However the fiber-local storage APIs do let you register a per-thread destructor. Even 
+// though we don't care about fibers, the API is strictly better and avoids the magic:
+// https://devblogs.microsoft.com/oldnewthing/20191011-00/?p=102989
```

Comment:
I think probably blink's way is better. The reason is fiber-local-storage is significantly slower than TLS.

Reply:
Yeah it does look like in single-threaded FlsGetValue is about 2x slower than TlsGetValue in the simple benchmarks I ran.

I think switching to TLS is fiddly enough that I'd rather land that in a future patch. It looks like libpas with fiber-local-storage is roughly on par with the system allocator for performance in the limited benchmarks I ran.